File: lib/contrek/finder/concurrent/cursor.rb
# frozen_string_literal: true

module Contrek
  module Concurrent
    class Cursor
      attr_reader :orphan_inners, :shapes_sequence

      def initialize(cluster:, shape:)
        @shapes_sequence = [shape]
        @cluster = cluster
        @outer_polyline = shape.outer_polyline
        @orphan_inners = []
      end

      def inspect
        self.class
      end

      # Given the initial polyline, draw its outer boundary, possibly extending into
      # adjacent polylines, and then connect them. At the end, @shapes_sequence
      # contains the merged polylines. Returns a new resulting polyline.
      def join_outers!
        seq_log = []

        outer_joined_polyline = Sequence.new
        traverse_outer(@outer_polyline.parts.first,
          seq_log,
          @shapes_sequence,
          outer_joined_polyline)
        outer_joined_polyline.pop! if outer_joined_polyline.head.payload == outer_joined_polyline.tail.payload &&
          @cluster.tiles.first.left? && @cluster.tiles.last.right?

        @shapes_sequence.each do |shape|
          shape.outer_polyline.turn_on(Polyline::TRACKED_OUTER)
          next if shape == @outer_polyline.shape
          @orphan_inners += shape.inner_polylines
          shape.clear_inner!
        end

        outer_joined_polyline
      end

      def join_inners!(treemap)
        return_inner_polylines = []
        shape_index = 0

        while shape_index < @shapes_sequence.size
          shape = @shapes_sequence[shape_index]
          polyline = shape.outer_polyline
          polyline.parts.each do |part|
            if part.innerable?
              all_parts = []
              tracked_end_points = []
              traverse_inner(part, all_parts, tracked_end_points)
              retme_sequence = Sequence.new
              all_parts.each do |part|
                part.touch!
                retme_sequence.append(part)
              end
              inner_polyline = InnerPolyline.new(sequence: retme_sequence)
              return_inner_polylines << inner_polyline
              mark_children(tracked_end_points, polyline, inner_polyline) if treemap
            end
          end
          shape_index += 1
        end
        return_inner_polylines
      end

      private

      def traverse_outer(start_part, all_parts, shapes_sequence, outer_joined_polyline)
        act_part = start_part

        while act_part
          last_part = all_parts.last
          all_parts << act_part if last_part != act_part
          jumped_to_new_part = false
          if act_part.is?(Part::EXCLUSIVE)
            return if act_part.size == 0
            while (position = act_part.next_position)
              outer_joined_polyline.add(position)
            end
          else
            while (new_position = act_part.iterator)
              if outer_joined_polyline.size > 1 &&
                  outer_joined_polyline.head.payload == new_position.payload &&
                  act_part == all_parts.first
                return
              end
              new_position.end_point.tracked_outer = true
              versus = act_part.versus
              part = new_position.end_point.queues.find do |p|
                (p.versus == -versus) && p.polyline.tile != act_part.polyline.tile
              end
              if part
                if all_parts[-2] != part
                  if !shapes_sequence.include?(part.polyline.shape)
                    shapes_sequence << part.polyline.shape
                  end
                  part.next_position(new_position)
                  act_part = part
                  jumped_to_new_part = true
                  break
                end
              end
              act_part.next_position
            end
          end
          next if jumped_to_new_part
          next_part = act_part.circular_next
          next_part.rewind!
          act_part = next_part
        end
      end

      # rubocop:disable Lint/NonLocalExitFromIterator
      def traverse_inner(act_part, all_parts, tracked_end_points)
        return if act_part == all_parts.first

        if act_part.innerable?
          all_parts << act_part
          while (act_part = act_part.circular_next)
            if act_part.innerable?
              all_parts << act_part
            else
              if act_part.head
                eligibles = act_part.head.end_point.queues.select { |p| p.polyline.tile != act_part.polyline.tile }
                eligibles.each do |dest_part|
                  dest_part_versus = dest_part.versus
                  next if dest_part_versus != 0 && dest_part_versus == act_part.versus

                  tracked_end_points << act_part.head.end_point
                  shape = dest_part.polyline.shape
                  if !dest_part.polyline.on?(Polyline::TRACKED_OUTER)
                    @shapes_sequence << shape
                    @orphan_inners += shape.inner_polylines
                    shape.clear_inner!
                  end
                  dest_part.polyline.turn_on(Polyline::TRACKED_OUTER)
                  if !dest_part.touched
                    dest_part.touch!
                    traverse_inner(dest_part.circular_next, all_parts, tracked_end_points)
                    return
                  end
                end
              end
              all_parts << act_part if act_part.is?(Part::SEAM)
            end
          end
        elsif act_part.next
          traverse_inner(act_part.next, all_parts, tracked_end_points)
        end
      end
      # rubocop:enable Lint/NonLocalExitFromIterator

      # finds each part (and relative polyline) inscribed between two end_points and sets the
      # founded inner_polyline which be later used to define in which parent hole is placed.
      def mark_children(end_points, outer_polyline, inner_polyline)
        end_points.each_slice(2) do |a, b|
          range = [a.position[:y], b.position[:y]].sort
          (range[0] + 1).upto(range[1] - 1) do |y|
            if (end_point = @cluster.hub.payloads[y])
              end_point.queues.each do |part|
                if part.polyline != outer_polyline
                  part.polyline.inside_inner_polyline = inner_polyline
                end
              end
            end
          end
        end
      end
    end
  end
end

File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/concurrent/Cursor.cpp
/*
 * Cursor.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */
#include <algorithm>
#include <vector>
#include <utility>
#include <unordered_set>
#include "Cursor.h"
#include "Polyline.h"
#include "InnerPolyline.h"

Cursor::Cursor(Cluster& cluster, Shape* shape)
: cluster(cluster), shape(shape) {
}

Sequence* Cursor::join_outers()
{ Polyline* outer_polyline = shape->outer_polyline;
  this->shapes_sequence_.push_back(this->shape);
  this->shapes_sequence_lookup.insert(this->shape);
  Sequence* outer_joined_polyline = outer_polyline->tile->shapes_pool->acquire_sequence();
  std::vector<Part*> all_parts;
  this->traverse_outer(outer_polyline->parts().front(),
                       all_parts,
                       this->shapes_sequence_,
                       outer_joined_polyline);

  if (*outer_joined_polyline->head->payload == *outer_joined_polyline->tail->payload &&
      this->cluster.tiles().front()->left() &&
      this->cluster.tiles().back()->right()) outer_joined_polyline->pop();

  for (Shape* shape : shapes_sequence_) {
    shape->outer_polyline->turn_on(Polyline::TRACKED_OUTER);
    if (shape == outer_polyline->shape) {
      continue;
    }
    orphan_inners_.insert(
        orphan_inners_.end(),
        shape->inner_polylines.begin(),
        shape->inner_polylines.end());
    shape->clear_inner();
  }
  return(outer_joined_polyline);
}

void Cursor::traverse_outer(Part* act_part,
                            std::vector<Part*>& all_parts,
                            std::vector<Shape*>& shapes_sequence,
                            Sequence* outer_joined_polyline) {
  while (act_part != nullptr) {
    Part* last_part = (!all_parts.empty()) ? all_parts.back() : nullptr;
    if (all_parts.empty() || last_part != act_part) {
      all_parts.push_back(act_part);
    }
    bool jumped_to_new_part = false;
    if (act_part->is(Part::EXCLUSIVE)) {
      if (act_part->size == 0) return;

      while (Position *position = act_part->next_position(nullptr)) {
        if (outer_joined_polyline->size > 1 &&
          outer_joined_polyline->head->payload == position->payload &&
          act_part == all_parts.front()) {
          return;
        }
        outer_joined_polyline->add(position);
      }
    } else {
      if (act_part->dead_end &&
         all_parts.size() > 1 &&
         last_part->is(Part::SEAM) &&
         last_part->polyline() == act_part->polyline()) return;
      while (Position *new_position = static_cast<Position*>(act_part->iterator())) {
        if (outer_joined_polyline->size > 1 &&
          outer_joined_polyline->head->payload == new_position->payload &&
          act_part == all_parts.front()) {
          return;
        }
        this->cluster.positions_pool.emplace_back(this->cluster.hub(), new_position->payload);
        outer_joined_polyline->add(&this->cluster.positions_pool.back());
        new_position->end_point()->tracked_outer = true;
        int versus = act_part->versus();
        auto& q_set = new_position->end_point()->queues();
        auto it = std::find_if(q_set.begin(), q_set.end(), [&](Queueable<Point>* q) {
          Part* p = static_cast<Part*>(q);
          return p->versus() == -versus && p->polyline()->tile != act_part->polyline()->tile;
        });

        Part* part = nullptr;
        if (it != q_set.end()) {
          part = static_cast<Part*>(*it);
        }
        if (part) {
          const auto n = all_parts.size();
          Part *last_last_part = n >= 2 ? all_parts[n - 2] : nullptr;
          if (last_last_part != part) {
            if (n >= 2) {
              bool all_seam = true;
              for (std::size_t i = all_parts.size() - 2; i < all_parts.size(); ++i) {
                if (all_parts[i]->type != Part::SEAM) {
                  all_seam = false;
                  break;
                }
              }
              if (all_seam) break;
            }
            if (shapes_sequence_lookup.insert(part->polyline()->shape).second) {
              shapes_sequence.push_back(part->polyline()->shape);
            }
            part->next_position(new_position);
            part->dead_end = true;
            act_part = part;
            jumped_to_new_part = true;
            break;
          }
        }
        if (!jumped_to_new_part) {
          act_part->next_position(nullptr);
        }
      }
    }
    if (jumped_to_new_part) continue;
    Part* next_part = act_part->circular_next;
    next_part->rewind();
    act_part = next_part;
  }
}

std::vector<InnerPolyline*> Cursor::join_inners(Sequence* outer_seq) {
  std::vector<InnerPolyline*> return_inner_polylines;
  std::vector<Shape*> processing_queue = shapes_sequence_;
  for (size_t i = 0; i < shapes_sequence_.size(); ++i)
  { Shape* shape = shapes_sequence_[i];
    Polyline* polyline = shape->outer_polyline;
    for (Part* part : polyline->parts())
    { if (part->innerable())
      { std::vector<Part*> all_parts;
        Bounds bounds{
          .min = polyline->max_y(),
          .max = 0
        };
        traverse_inner(part, all_parts, bounds);
        Sequence* retme_sequence = shape->outer_polyline->tile->shapes_pool->acquire_sequence();
        for (Part* part : all_parts)
        { part->touch();
          retme_sequence->move_from(*part, [&](QNode<Point>* pos) -> bool {
          Position *position = static_cast<Position*>(pos);
          if (part->is(Part::ADDED) &&
             !(position->payload->y >= bounds.min &&
              position->payload->y <= bounds.max)) {
            return(false);
          }
            return(!(polyline->tile->tg_border(*position->payload) && position->end_point()->tracked_outer));
          });
        }
        if (retme_sequence->is_not_vertical()) {
          return_inner_polylines.push_back(polyline->tile->shapes_pool->acquire_inner_polyline(retme_sequence));
        }
      }
    }
  }
  return(return_inner_polylines);
}

void Cursor::traverse_inner(Part* act_part, std::vector<Part*>& all_parts, Bounds& bounds) {
  PartPool& pool = act_part->polyline()->tile->cluster->parts_pool;
  while (act_part != nullptr) {
    if (!all_parts.empty() && act_part == all_parts.front()) return;
    if (act_part->size > 0) {
      auto points = act_part->to_vector();
      auto [min_it, max_it] = std::minmax_element(
        points.begin(),
        points.end(),
        [](Point* a, Point* b) { return a->y < b->y; });
      bounds.min = std::min(bounds.min, (*min_it)->y);
      bounds.max = std::max(bounds.max, (*max_it)->y);
    }
    if (act_part->innerable()) {
      all_parts.push_back(act_part);
      bool jumped = false;
      while (act_part = act_part->circular_next) {
        if (act_part->innerable()) {
            all_parts.push_back(act_part);
        } else {
          if (act_part->head)
          { for (auto dest_part_p : static_cast<Position*>(act_part->head)->end_point()->queues()) {
              Part* dest_part = static_cast<Part*>(dest_part_p);
              if (dest_part->polyline()->tile == act_part->polyline()->tile) {
                continue;
              }
              int dest_part_versus = dest_part->versus();
              if (dest_part_versus != 0 && dest_part_versus == act_part->versus()) continue;
              std::vector<EndPoint*> link_seq = dest_part->continuum_to(*act_part);
              if (!link_seq.empty()) {
                Part* ins_part = pool.acquire(Part::ADDED, act_part->polyline());
                for (EndPoint* pos : link_seq) {
                  this->cluster.positions_pool.emplace_back(pos);
                  ins_part->add(&this->cluster.positions_pool.back());
                }
                all_parts.push_back(ins_part);
              }
              Shape* shape = dest_part->polyline()->shape;
              if (!dest_part->polyline()->is_on(Polyline::TRACKED_OUTER))
              { shapes_sequence_.push_back(shape);
                orphan_inners_.insert(
                  orphan_inners_.end(),
                  shape->inner_polylines.begin(),
                  shape->inner_polylines.end());
              }
              dest_part->polyline()->turn_on(Polyline::TRACKED_OUTER);
              if (!dest_part->touched()) {
                dest_part->touch();

                act_part = dest_part->circular_next;
                jumped = true;
                break;
              }
            }
            if (jumped) break;
          }
          if (act_part->is(Part::SEAM)) {
            all_parts.push_back(act_part);
          }
        }
      }
      if (jumped) continue;
      break;
    } else if (act_part->next) {
      act_part = act_part->next;
      continue;
    }
    else break;
  }
}
